app/uk/gov/hmrc/gform/models/ObligationValidator.scala

```diff
@@ -16,50 +16,33 @@
-
-trait ObligationValidator {
-
-  def validate(data: Data, cachedObligation: Obligation, taxResponse: TaxResponse): TaxSelectionNavigator = {
-    val obligationsMatch = obligationsMatchTaxResponseObligations(cachedObligation, taxResponse)
-    val stillAvailable = selectedPeriodStillAvailable(data, taxResponse)
-
-    (obligationsMatch, stillAvailable) match {
-      case (GoBackToTaxPeriodSelection, DoNotGoBackToTaxPeriodSelection) =>
-        goBackIf(desHasMore(cachedObligation, taxResponse.obligation))
-      case (GoBackToTaxPeriodSelection, GoBackToTaxPeriodSelection) =>
-        GoBackToTaxPeriodSelection
-      case (DoNotGoBackToTaxPeriodSelection, _) =>
-        DoNotGoBackToTaxPeriodSelection
+import cats.data.NonEmptyList
+import uk.gov.hmrc.gform.sharedmodel.form.FormDataRecalculated
+import uk.gov.hmrc.gform.sharedmodel.{ Obligations, RetrievedObligations, TaxResponse }
+
+trait ObligationValidator extends TaxSelectionNavigator {
+
+  def validateWithDes(
+    formDataRecalculated: FormDataRecalculated,
+    cachedObligation: Obligations,
+    desObligation: Obligations,
+    clearTaxResponses: FormDataRecalculated => FormDataRecalculated): FormDataRecalculated =
+    (cachedObligation, desObligation) match {
+      case (RetrievedObligations(obligation), RetrievedObligations(responseObligation))
+          if mayClear(formDataRecalculated, obligation, responseObligation) =>
+        clearTaxResponses(formDataRecalculated)
+      case _ => formDataRecalculated
     }
-  }
-
-  private def obligationsMatchTaxResponseObligations(
-    cachedObligation: Obligation,
-    taxResponse: TaxResponse): TaxSelectionNavigator =
-    if (cachedObligation.obligations == taxResponse.obligation.obligations) DoNotGoBackToTaxPeriodSelection
-    else GoBackToTaxPeriodSelection
-
-  private def selectedPeriodStillAvailable(data: Data, taxResponse: TaxResponse): TaxSelectionNavigator = {
-    val desPeriods: List[String] = taxResponse.obligation.obligations.flatMap(_.obligationDetails).map(_.periodKey)
-    (for {
-      maybePeriod <- data.get(taxResponse.id.recalculatedTaxPeriodKey.fcId)
-      periodValue <- maybePeriod.headOption
-    } yield periodValue).fold[TaxSelectionNavigator](GoBackToTaxPeriodSelection) { period =>
-      goBackIf(!desPeriods.contains(period))
-    }
-  }
-
-  private def desHasMore(dataObligation: Obligation, desObligation: Obligation): Boolean =
-    desObligation.obligations.flatMap(_.obligationDetails).size > dataObligation.obligations
-      .flatMap(_.obligationDetails)
-      .size
 
-  private val goBackIf: Boolean => TaxSelectionNavigator =
-    predicate => if (predicate) GoBackToTaxPeriodSelection else DoNotGoBackToTaxPeriodSelection
+  private def mayClear(
+    formDataRecalculated: FormDataRecalculated,
+    cachedObligation: NonEmptyList[TaxResponse],
+    responseObligation: NonEmptyList[TaxResponse]): Boolean =
+    cachedObligation.toList
+      .zip(responseObligation.toList)
```

Comment:
`cachedObligation` and `responseObligation` are not guaranteed to be in the same order unfortunately.

The reason is that they are dependent on a dependency graph construction which is not stable. For example [here](https://github.com/hmrc/gform-frontend/blob/master/test/uk/gov/hmrc/gform/graph/DependencyGraphSpec.scala#L242) we are sorting the dependency graph explicitly for convenience in our tests.